packages/Discretization/test/PointCloudProblemGenerator/ExodusAsciiGenerator_impl.hpp

```diff
@@ -0,0 +1,312 @@
+    const int dim,
+    const std::string& coord_file,
+    Kokkos::View<Coordinate**,Kokkos::LayoutLeft,Device>& device_coords )
+{
+    // Get the node data.
+    Kokkos::View<Coordinate**,Kokkos::LayoutLeft,Kokkos::Serial> host_coords;
+    Kokkos::View<GlobalOrdinal*,Kokkos::LayoutLeft,Kokkos::Serial> host_gids;
+    getNodeDataFromFile( coord_file, host_coords, host_gids );
+
+    // Partition the coordinates in the right direction. Figure out the min
+    // and max coordinates.
+    Coordinate dim_max, dim_min;
+    dimensionMinAndMax( dim, host_coords, dim_min, dim_max );
+
+    // Build a communication plan.
+    int num_node = host_coords.extent( 0 );
+    Teuchos::Array<int> export_ranks( num_node );
+    double dim_frac = 0.0;
+    for ( int n = 0; n < num_node; ++n )
+    {
+        dim_frac = (host_coords(n,dim)-dim_min) / (dim_max-dim_min);
+        export_ranks[n] = (1.0 == dim_frac)
+                          ? _comm->getSize() - 1
+                          : std::floor( dim_frac * _comm->getSize() );
+    }
+    Tpetra::Distributor distributor( _comm );
+    int num_node_import = distributor.createFromSends( export_ranks() );
+
+    // Send the sources to their new destination.
+    Kokkos::View<Coordinate**,Kokkos::LayoutLeft,Kokkos::Serial> import_coords(
+        "import_coords", num_node_import, 3 );
+    for ( int d = 0; d < 3; ++d )
+    {
+        distributor.doPostsAndWaits(
+            Kokkos::subview(host_coords,Kokkos::ALL,d),
+            1,
+            Kokkos::subview(import_coords,Kokkos::ALL,d) );
+    }
+
+    // Move the coordinates to the device.
+    device_coords =
+        Kokkos::View<Coordinate**,Kokkos::LayoutLeft,Device>( "device_coords", num_node_import, 3 );
+    Kokkos::deep_copy( device_coords, import_coords );
+}
+
+//---------------------------------------------------------------------------//
+// Partition a point cloud in a given dimension with ghosted connectivity
+// mapping.
+template<class SourceDevice, class TargetDevice>
+template<class Device>
+void ExodusAsciiGenerator<SourceDevice,TargetDevice>::partitionGhostedConnectivity(
+    const int dim,
+    const std::string& coord_file,
+    const std::string& connectivity_file,
+    Kokkos::View<Coordinate**,Kokkos::LayoutLeft,Device>& device_coords,
+    Kokkos::View<GlobalOrdinal*,Kokkos::LayoutLeft,Device>& device_gids )
+{
+    // Get the node data.
+    Kokkos::View<Coordinate**,Kokkos::LayoutLeft,Kokkos::Serial> host_coords;
+    Kokkos::View<GlobalOrdinal*,Kokkos::LayoutLeft,Kokkos::Serial> host_gids;
+    getNodeDataFromFile( coord_file, host_coords, host_gids );
+
+    // Partition the source coordinates in the dimension. Figure out the min
+    // and max coordinates.
+    Coordinate dim_max, dim_min;
+    dimensionMinAndMax( dim, host_coords, dim_min, dim_max );
+
+    // Load the connectivity file.
+    std::ifstream conn_file;
+    conn_file.open( connectivity_file );
```

Comment:
You want every processor to open the file?

Reply:
good catch